pkg/server/endpoints/eventTracker_test.go

```diff
@@ -0,0 +1,1120 @@
+			expectedBoundaries:   []uint{0},
+		},
+		{
+			name:        "polling twice, both initial",
+			pollPeriods: 2,
+			boundaries:  []uint{},
+
+			expectedInitialPolls: 2,
+			expectedPollPeriods:  2,
+			expectedPolls:        2,
+			expectedBoundaries:   []uint{},
+		},
+		{
+			name:        "polling twice, once initial, once in one bucket boundary",
+			pollPeriods: 2,
+			boundaries:  []uint{1},
+
+			expectedInitialPolls: 1,
+			expectedPollPeriods:  2,
+			expectedPolls:        2,
+			expectedBoundaries:   []uint{1},
+		},
+		{
+			name:        "polling once, in two bucket boundary",
+			pollPeriods: 2,
+			boundaries:  []uint{0},
+
+			expectedInitialPolls: 0,
+			expectedPollPeriods:  2,
+			expectedPolls:        1,
+			expectedBoundaries:   []uint{0},
+		},
+		{
+			name:        "polling once, in three bucket boundary",
+			pollPeriods: 3,
+			boundaries:  []uint{0},
+
+			expectedInitialPolls: 0,
+			expectedPollPeriods:  3,
+			expectedPolls:        1,
+			expectedBoundaries:   []uint{0},
+		},
+		{
+			name:        "polling six times in exponential backoff",
+			pollPeriods: 120,
+			boundaries:  []uint{0, 2, 6, 14, 30, 62},
+
+			expectedInitialPolls: 0,
+			expectedPollPeriods:  120,
+			expectedPolls:        6,
+			expectedBoundaries:   []uint{0, 2, 6, 14, 30, 62},
+		},
+		{
+			name:        "distributed linear polling for a while, then exponential",
+			pollPeriods: 600,
+			boundaries:  []uint{0, 10, 20, 30, 40, 50, 60, 120, 240, 480},
+
+			expectedInitialPolls: 0,
+			expectedPollPeriods:  600,
+			expectedPolls:        10,
+			expectedBoundaries:   []uint{0, 10, 20, 30, 40, 50, 60, 120, 240, 480},
+		},
+		{
+			name:        "clip boundaries outside of poll periods",
+			pollPeriods: 600,
+			boundaries:  []uint{0, 10, 20, 30, 40, 50, 60, 120, 240, 480, 9600},
+
+			expectedInitialPolls: 0,
+			expectedPollPeriods:  600,
+			expectedPolls:        10,
+			expectedBoundaries:   []uint{0, 10, 20, 30, 40, 50, 60, 120, 240, 480},
+		},
+		{
+			name:        "order of boundaries doesn't matter",
+			pollPeriods: 600,
+			boundaries:  []uint{240, 480, 9600, 0, 10, 50, 60, 120, 20, 30, 40},
+
+			expectedInitialPolls: 0,
+			expectedPollPeriods:  600,
+			expectedPolls:        10,
+			expectedBoundaries:   []uint{0, 10, 20, 30, 40, 50, 60, 120, 240, 480},
+		},
+		{
+			name:        "duplicate boundaries are collapsed",
+			pollPeriods: 600,
+			boundaries:  []uint{0, 10, 10, 10, 20, 30, 40, 50, 60, 60, 120, 240, 480, 240, 9600},
+
+			expectedInitialPolls: 0,
+			expectedPollPeriods:  600,
+			expectedPolls:        10,
+			expectedBoundaries:   []uint{0, 10, 20, 30, 40, 50, 60, 120, 240, 480},
+		},
+	} {
+		tt := tt
+		t.Run(tt.name, func(t *testing.T) {
+			eventTracker := endpoints.NewEventTracker(tt.pollPeriods, tt.boundaries)
+
+			require.Equal(t, tt.expectedPollPeriods, eventTracker.PollPeriods(), "expcting %d poll periods; but, %d poll periods reported", eventTracker.PollPeriods(), tt.expectedPollPeriods)
```

Comment:
hypernit, and maybe it's just my style- in tests I find using `require` aggressively may hide signal from code maintainers. I only use it if there would later be a panic (such as in a nil pointer) or there's otherwise no point in continuing the test (like if a dependency wasn't created correctly). in most other cases, `assert` will let a contributor get all relevant signal on what failed. as an example here, it could be interesting for a contributor to know that they broke the code in such a way that poll periods failed *and* boundaries failed but not expectedpools.

Reply:
This is interesting, and I just read up on the differences between `require` and `assert`.  That said, there's no acceptable test in the suite that can fail and the product still be deemed "ok to use" because any failure would mean a runtime deviation between the cache and the database.

If it's a nit that keeps bothering you I can update the code, but any failure probably should just hurry through to the module failing unit test.

Reply:
not at all necessary to update- for me nits are always optional but just something to point out :)